[#8726] Port twisted.protocols.memcache to Python 3

charmander opened the pull request:
https://twistedmatrix.com/trac/ticket/8726

Includes some questionable `.decode("ascii")`s. Is them raising an exception something to worry about?

charmander committed:
Port twisted.protocols.memcache to Python 3

Includes some questionable `.decode("ascii")`s. Is them raising
an exception something to worry about?

rodrigc commented on twisted/protocols/memcache.py:
```diff
@@ -622,7 +627,7 @@ def getMultiple(self, keys, withIdentifier=False):
         issuing C{checkAndSet} update later, using the corresponding method.
 
         @param keys: The keys to retrieve.
-        @type keys: C{list} of C{str}
```
Since you are taking a pass through this file can you change C{int}, C{str}, C{list}, C{bytes} to L{int}, L{str}, L{list}, L{bytes}?

According to the coding standard, basic Python types need to be wrapped in L{} so that they link back to the Python documentation:

https://twistedmatrix.com/documents/current/core/development/policy/coding-standard.html#docstrings

charmander committed:
Use L{} to link to built-in Python types

rodrigc commented on twisted/protocols/memcache.py:
```diff
@@ -25,12 +25,15 @@ def doSomething(proto):
 from twisted.protocols.policies import TimeoutMixin
 from twisted.internet.defer import Deferred, fail, TimeoutError
 from twisted.python import log
+from twisted.python.compat import iteritems
```
Can you import nativeString and networkString from twisted.python.compat?
In this file, for all changes where you did  foo.encode("ascii"), replace with networkString(foo),
and for all cases of bar.decode("ascii"), replace with nativeString(bar).

This will keep things compatible on Python 2 and Python 3

rodrigc commented on twisted/protocols/memcache.py:
```diff
@@ -343,19 +347,19 @@ def lineReceived(self, line):
+        token = line.split(b" ", 1)[0]
         # First manage standard commands without space
-        cmd = getattr(self, "cmd_%s" % (token,), None)
+        cmd = getattr(self, u"cmd_" + token.decode("ascii"), None)
```
this is better as cmd = getattr(self, "cmd_" + nativeString(token), None)
No need to convert to unicode here, since this is a behavior change on Python 2.

charmander committed:
Raise error on unexpected memcached END response

charmander committed:
Make use of nativeString and networkString

Maintains backwards-compatibility with regards to text types
used in getattr() and error messages.

rodrigc commented:
Thanks for the updates.  This is almost good to go in.
If I look at this coverage report, on line 242:
https://codecov.io/gh/twisted/twisted/compare/3e488bca70159bd8d7f5a62dc55e0ab9879f229b...5288541dce2e97b5b323ccdc0267bbe52050aaa5?src=pr#747769737465642F70726F746F636F6C732F6D656D63616368652E7079-242

It looks like the existing tests don't cover the **b'stats'** value.  Can you extend the existing tests by an additional test that tests for this value?

charmander commented:
@rodrigc 554fdf86df6a4ce2060c6309183d30a4c7a77b24 addressed that, but it looks like Codecov didn’t pick it up for some reason. (There are only two checks listed instead of the earlier four.) Should I merge trunk to try to trigger it?

hawkowl commented on twisted/protocols/memcache.py:
```diff
@@ -299,20 +305,22 @@ def cmd_ERROR(self):
 
 
-    def cmd_CLIENT_ERROR(self, errText):
+    def cmd_CLIENT_ERROR(self, errBytes):
```
I believe this breaks our backwards compatibility policy, because it's renaming parts of our public API :(

It does make it clearer, though, so I believe that maybe something like:

```
def somefunc(self, errBytes=None, errText=None):
    if errText:
        raise DeprecationWarning("Passing 'errText' as a kwarg to somefunc is deprecated since Twisted 16.4.0")
    if errText and errBytes is None:
        errBytes = errText
    elif errText and errBytes:
        raise ValueError("Please specify only only one argument (errBytes).")
```

or something...

cc @rodrigc , @glyph

charmander replied:
I didn’t notice it was public – whoops! In that case, the bytes still have to represent text, so maybe it would be best to just keep the name as-is.

charmander committed:
Restore errText as a public parameter name

adiroiban commented:
Changes looks good. Thanks!

Please include twisted/test/test_memcache.py to dist3.py so that we will run the test on python3

decode('ascii') on random input can be a problem as if we receive non 7bit ascii bytes decode will break... is better to just user repr() 

---

codecov.io might be slow to report the coverage

charmander committed:
Add twisted.test.test_memcache to Python 3 tests

rodrigc commented:
@charmander Your build is failing on Python 3.3 and Python 3.4, like here: https://travis-ci.org/twisted/twisted/jobs/150782318#L2477

Python 3.3 and 3.4 did not support string formatting of bytes.  That only came into Python 3.5:
https://www.python.org/dev/peps/pep-0461/

Can you change your code so that it works on Python 3.3 and 3.4?  Twisted needs to work on those versions also.

charmander commented:
Ah, I took that into account for `twisted.protocol.memcache` but missed it in the tests. Sorry!

charmander committed:
Remove %-formatting of bytes in memcache tests

hawkowl commented on twisted/protocols/memcache.py:
```diff
@@ -411,12 +419,12 @@ def _incrdecr(self, cmd, key, val):
         if len(key) > self.MAX_KEY_LENGTH:
             return fail(ClientError("Key too long"))
-        fullcmd = "%s %s %d" % (cmd, key, int(val))
+        fullcmd = b" ".join([cmd, key, networkString(str(int(val)))])
```
`twisted.python.compat.intToBytes` would be nicer here.

charmander replied:
It would. Thanks!

charmander committed:
Use intToBytes instead of networkString on str

hawkowl commented:
Thanks for this @charmander .

re: `encode('ascii')` (and by extension `networkString`, which does that internally), that is kind of the default assumption of Python 2/3 if you pass it Unicode, but I think random 8-bit trash is actually allowed: https://github.com/memcached/memcached/blob/master/doc/protocol.txt#L495

I am not sure what this really means for this branch, I think we ought to allow binary trash in keys, since it used to work (and this branch might cause working Python 2 apps to stop working). So, I'll have a look through and point out some places where we might need to fix it.

hawkowl commented on twisted/protocols/memcache.py:
```diff
@@ -303,7 +309,8 @@ def cmd_CLIENT_ERROR(self, errText):
         An invalid input as been sent.
         """
-        log.err("Invalid input: %s" % (errText,))
+        errText = nativeString(errText)
```
This should `repr()` instead -- the err will be `b''` prefixed on Python 3, but that's fine, as it means binary garbage coming in won't cause this to fail.

hawkowl commented on twisted/protocols/memcache.py:
```diff
@@ -312,7 +319,8 @@ def cmd_SERVER_ERROR(self, errText):
         An error has happened server-side.
         """
-        log.err("Server error: %s" % (errText,))
+        errText = nativeString(errText)
```
Likewise with the repr comment.

hawkowl commented on twisted/test/test_memcache.py:
```diff
@@ -5,6 +5,8 @@
 Test the memcache client protocol.
 """
 
```
This needs a `__future__` import, like the implementation function.

hawkowl commented on twisted/test/test_memcache.py:
```diff
@@ -278,11 +281,11 @@ def test_invalidGetResponse(self):
         self.assertRaises(
             RuntimeError, self.proto.dataReceived,
-            "VALUE bar 0 %s\r\n%s\r\nEND\r\n" % (len(s), s))
+            networkString("VALUE bar 0 %d\r\n%s\r\nEND\r\n" % (len(s), s)))
```
I would much prefer this and the next test to use `intToBytes` vs `networkString`.

hawkowl replied:
Actually, other tests hardcode the length. I'd be fine with that too, then it's consistent.

charmander commented:
This shouldn’t attempt to decode keys anywhere… unless they’re included raw in some error messages. That doesn’t seem to be the case judging by a quick search of memcached, but I don’t know about past versions.

hawkowl commented:
So I've read through and am confident that no encoding/decoding is done on keys, so that shouldn't be a worry. I raised one or two places where encoding problems may occur,  depending on what the server sends (although it shouldn't send back binary trash, but I'd prefer to be cautious here)...

@charmander , thanks again! After those little things, I think we are good to land!

charmander committed:
Hardcode all memcached test responses

charmander committed:
repr memcached error messages instead of decoding them

Just in case. Also `__future__`ize the tests.

charmander committed:
MIT-license changes

charmander committed:
Allow keys to be any iterable in multiple get

The original implementation didn’t allow non-reusable iterables, but it
did work with tuples, which are rather likely to be in use! Fixes that
compatibility issue.

hawkowl commented:
I'm gonna merge this. Thanks again @charmander !
